File: TunnelSimulator.UI/src/services/tunnel/tunnel.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { TunnelStatusResponseModel } from './models/tunnel-status-response-model';
import { TunnelSetCalamityRequestModel } from './models/tunnel-set-calamity-request-model';

@Injectable({
  providedIn: 'root'
})
export class TunnelService {
  private apiUrl = '/api/tunnel';

  constructor(private http: HttpClient) { }

  getTunnelStatus(): Observable<TunnelStatusResponseModel> {
    return this.http.get<TunnelStatusResponseModel>(`${this.apiUrl}`);
  }

  setCalamity(tunnelTube: number, requestModel: TunnelSetCalamityRequestModel) {
    return this.http.post(`${this.apiUrl}/${tunnelTube}/Calamity`, requestModel);
  }
}

File: TunnelSimulator.API/TunnelSimulator.API/Services/Tunnel/TunnelService.cs
using TunnelSimulator.API.Services.Tunnel.Models;
using TunnelSimulator.API.Services.Websocket;

namespace TunnelSimulator.API.Services.Tunnel;

public class TunnelService(TunnelWebSocketHandler tunnelWebSocketHandler) : ITunnelService
{
    public static TunnelStatusResponseModel TunnelStatus = new TunnelStatusResponseModel()
    {
        Tubes =
        [
            new TunnelTubeStatusResponseModel()
            {
                IsCalamity = false,
            },
            new TunnelTubeStatusResponseModel()
            {
                IsCalamity = false,
            }
        ]
    };

    public TunnelStatusResponseModel GetTunnelStatus()
    {
        return TunnelStatus;
    }

    public async Task SetCalamity(int tunnelTube, TunnelSetCalamityRequestModel model)
    {
        TunnelStatus.Tubes[tunnelTube].IsCalamity = model.IsCalamity;
        await tunnelWebSocketHandler.TunnelStatusUpdated();
    }
}
